### PR TITLE
CI: split Nix cache as separate save and restore cache

### LINKED ISSUE #6522: CI - Split the nix cache composite action to a restore and save action
We have a [composite action that caches our nix dependencies and binary](https://github.com/build-trust/ockam/blob/353f4be70d0d9455437ce20044e6208e268d306b/.github/actions/nix_installer_and_cache/action.yml#L17-L23), we should split this action to support [saving](https://github.com/actions/cache/blob/main/save/README.md) and [restoring](https://github.com/actions/cache/blob/main/restore/README.md) distinctively, we should only save cache if and only if the branch is from the default (develop) branch so that we can [cache efficiently](https://docs.github.com/en/actions/using-workflows/caching-dependencies-to-speed-up-workflows), to allow this

- Only use [the actions/cache action](https://github.com/build-trust/ockam/blob/353f4be70d0d9455437ce20044e6208e268d306b/.github/actions/nix_installer_and_cache/action.yml#L17-L23) (which saves and restore cache) only when the workflow is run as a cron job by adding a [conditional](https://docs.github.com/en/actions/using-jobs/using-conditions-to-control-job-execution) to ensure that the [event_name that triggered the workflow](https://docs.github.com/en/actions/learn-github-actions/contexts#github-context) is a [schedule](https://docs.github.com/en/actions/using-workflows/events-that-trigger-workflows#schedule) `${{ github.event_name == 'schedule' }}`
- If worklfow is not run from a scheduler, we should use only [restore cache action](https://github.com/actions/cache/blob/main/restore/README.md)

---
<i>

We love helping new contributors! ❤️
If you have questions or need help as you explore, please join us on [Discord](https://discord.gg/bewvnm6zqS). If you're looking for other issues to contribute to, please checkout our [good first issues](https://github.com/build-trust/ockam/labels/good%20first%20issue).

</i>

### PR DESCRIPTION
**Pull request by nabeel001**
## Current behavior

Closes #6522 

## Proposed changes

As described in the issue the Nix cache needs to be split into separate save and restore cache.

## Checks

- [x] All commits in this Pull Request are [signed](https://docs.github.com/en/authentication/managing-commit-signature-verification/signing-commits) and Verified by Github.
- [x] All commits in this Pull Request follow the Ockam [commit message convention](https://github.com/build-trust/.github/blob/main/CONTRIBUTING.md#commit-messages).
- [ ] There are no Merge commits in this Pull Request. Ockam repo maintains a linear commit history. We merge Pull Requests by rebasing them onto the develop branch. Rebasing to the latest develop branch and force pushing to your Pull Request branch is okay.
- [x] I have read and accept the Ockam Community [Code of Conduct](https://github.com/build-trust/.github/blob/main/CODE_OF_CONDUCT.md).
- [x] I have read and accepted the Ockam [Contributor License Agreement](https://github.com/build-trust/.github/blob/main/CLA.md) by adding my Git/Github details in a row at the end of the [CONTRIBUTORS.csv](https://github.com/build-trust/ockam/blob/develop/.github/CONTRIBUTORS.csv) file in a separate pull request to the [build-trust/ockam](https://github.com/build-trust/ockam) repository. The easiest way to do this is to [edit the CONTRIBUTORS.csv](https://github.com/build-trust/ockam/edit/develop/.github/CONTRIBUTORS.csv) file in the github web UI and create a separate Pull Request, this will mark the commit as verified.

### CODE REVIEW

File: .github/actions/nix_installer_and_cache/action.yml [hunk 1]
```diff
@@ -13,38 +13,48 @@ outputs:
...
+        if: steps.nix-cache.outputs.cache-hit != 'true'
+        uses: DeterminateSystems/nix-installer-action@3ebd1aebb47f95493b62de6eec0cac3cd74e50a9
+
+      - name: restore cache
+        if: ${{ github.event_name != 'schedule' }}
+        uses: actions/cache/restore@704facf57e6136b1bc63b828d79edcd491f0ee84
+        with:
+          path: |
+            /tmp/nix-cache
+          key: nix-cache-${{ runner.os }}-${{ github.workflow }}-${{ github.job }}-${{ inputs.cache-unique-id }}-${{ hashFiles('**/flake.lock') }}
```
Comment (metaclips) [inline .github/actions/nix_installer_and_cache/action.yml]: Let's make this step the second step that runs if `restore and save cache` conditional step doesn't run

File: .github/actions/nix_installer_and_cache/action.yml [hunk 2]
```diff
@@ -13,38 +13,48 @@ outputs:
...
+          sudo mv /tmp/nix-cache/** /nix
+  
+          echo "/nix/var/nix/profiles/per-user/root/profile/bin" >> $GITHUB_PATH
+          echo "/nix/var/nix/profiles/default/bin" >> $GITHUB_PATH
+  
+          sudo chown -R "${USER}:" /nix
+  
+      - name: Install Nix
+        if: steps.nix-cache.outputs.cache-hit != 'true'
+        uses: DeterminateSystems/nix-installer-action@3ebd1aebb47f95493b62de6eec0cac3cd74e50a9
```
Comment (metaclips) [inline .github/actions/nix_installer_and_cache/action.yml]: We only run this conditional if `restore and save` action runs successfully, we should also run this action if the `restore cache` action runs too.

```yaml
- name: restore and save cache
   id: nix-restore-and-save
   if: ${{ github.event_name == 'schedule' }}
   .....

- name: restore cache
   id: nix-restore
   if: ${{ github.event_name != 'schedule' }}
   ......

- name: cache hit
   if: ${{ steps.nix-restore-and-save.outputs.cache-hit == 'true' || steps.nix-restore.outputs.cache-hit == 'true' }}
   ....
```

**Review by metaclips — CHANGES_REQUESTED** [review]
Way to go @nabeel001. Little changes left to get this PR merged

File: .github/actions/nix_installer_and_cache/action.yml [hunk 3]
```diff
@@ -13,38 +13,49 @@ outputs:
...
+  
+          sudo mv /tmp/nix-cache/** /nix
+  
+          echo "/nix/var/nix/profiles/per-user/root/profile/bin" >> $GITHUB_PATH
+          echo "/nix/var/nix/profiles/default/bin" >> $GITHUB_PATH
+  
+          sudo chown -R "${USER}:" /nix
+  
+      - name: Install Nix
+        if: steps.nix-cache.outputs.cache-hit != 'true'
```
Comment (metaclips) [inline .github/actions/nix_installer_and_cache/action.yml]: `if: ${{ steps.nix-restore-and-save.outputs.cache-hit == 'true' || steps.nix-restore.outputs.cache-hit == 'true' }}`

**Review by metaclips — CHANGES_REQUESTED** [review]
Last change

**Comment by nabeel001** [comment]
@metaclips, I have made the requested changes.

File: .github/actions/nix_installer_and_cache/action.yml (hunk 3 above)
Comment (nabeel001) [reply .github/actions/nix_installer_and_cache/action.yml]: oops, My bad I missed it.

File: .github/actions/nix_installer_and_cache/action.yml [hunk 4]
```diff
@@ -13,38 +13,49 @@ outputs:
...
-        echo "/nix/var/nix/profiles/default/bin" >> $GITHUB_PATH
-
-        sudo chown -R "${USER}:" /nix
-
-    - name: Install Nix
-      if: steps.nix-cache.outputs.cache-hit != 'true'
-      uses: DeterminateSystems/nix-installer-action@3ebd1aebb47f95493b62de6eec0cac3cd74e50a9
+jobs:
+  cache:
+    runs-on: ubuntu-latest
```
Comment (metaclips) [inline .github/actions/nix_installer_and_cache/action.yml]: Please revert this change. This action is a [composite action](https://docs.github.com/en/actions/creating-actions/creating-a-composite-action) not a job
```yaml
runs:
  using: composite
```

Comment (nabeel001) [reply .github/actions/nix_installer_and_cache/action.yml]: Done

File: .github/CONTRIBUTORS.csv [hunk 5]
```diff
@@ -320,3 +320,4 @@ yes,Suraj-kumar00,Suraj kumar,<suraj.ukumar.p@gmail.com> <123288511+Suraj-kumar0
 yes,Lucif3r-in,Ashutosh Rath,<ashutosh123rath@gmail.com> <85403534+Lucif3r-in@users.noreply.github.com>
 yes,Lumexralph,Olumide Ogundele,<olumideralph@gmail.com>
 yes,utkarsh1042005,Utkarsh,<kumarutkarsh563@gmail.com> <146640299+utkarsh1042005@users.noreply.github.com>
+yes,nabeel001,Nabeel Hussain,<68153692+nabeel001@users.noreply.github.com>
```
Comment (metaclips) [inline .github/CONTRIBUTORS.csv]: @nabeel001 you'll need to create a different pull-request that signs your CLA, then we can merge this PR. Everyother thing looks fine

Comment (nabeel001) [reply .github/CONTRIBUTORS.csv]: ok

**Comment by nabeel001** [comment]
@metaclips @mrinalwadhwa I have added my name to CONTRIBUTORS.md through #6616

**Review by metaclips — APPROVED** [review]

**Review by davide-baldo — APPROVED** [review]
